Update color space if HDR is toggled in settings

jpark37 opened the pull request:
Can't use GetContainingOutput because it returns stale info.

jpark37 commented:
I see there are many DeviceResources.cpp out there, but I figured you guys would know how to propagate that better than me.

Assuming I didn't make a mistake in testing, I did verify that SimpleHDR_PC didn't respect HDR toggle without the changes, still didn't respect HDR toggle without the GetContainingOutput/factory replacement, but did work with all the changes.

walbourn commented:
Thanks for the feedback.

So the basic issue here is GetContainingOutput in recent Windows 10 versions doesn't do the right thing here...

Also, I forgot to call UpdateColorSpace after refreshing the factory. I tested moving between an HDR monitor and a non-HDR monitor, but didn't every try the toggle.

walbourn commented:
And yes, I'll need to propagate this fix to a lot of places.  Much appreciated though :)

walbourn commented:
Note that for UWP I can't use ``MonitorFromWindow`` since it's not in the UWP partition. That's why I used GetContainingOutput in the first place.

jpark37 commented:
Fixed the leaks, and changed the loops to match existing code pattern.

> So the basic issue here is GetContainingOutput in recent Windows 10 versions doesn't do the right thing here...

I'm not sure how it can since the swap chain is tied to a non-current factory.

> Note that for UWP I can't use MonitorFromWindow since it's not in the UWP partition. That's why I used GetContainingOutput in the first place.

I dunno how UWP security works, but when I run SimpleHDR_UWP and step into GetContainingOutput, I see this call stack.

```
>	user32.dll!MonitorFromWindow()	Unknown
 	dxgi.dll!GetContainedOutputHwnd(struct HWND__ *,class ATL::CComPtr<class CDXGIFactory> const &,class CDXGIAdapter *,struct IDXGIOutput * *)	Unknown
 	dxgi.dll!CDXGISwapChain::GetContainingOutputImpl(struct IDXGIOutput * *,bool)	Unknown
 	dxgi.dll!CDXGISwapChain::GetContainingOutput(struct IDXGIOutput * *)	Unknown
 	SimpleHDR_UWP.exe!DX::DeviceResources::UpdateColorSpace() Line 692	C++
```

This DLL is allowed to call the function, but the application isn't?

jpark37 commented:
Removed stray variable, and switched to MONITOR_DEFAULTTONEAREST to match DXGI implementation.

jpark37 commented:
Aren't UWP apps supposed to call `DisplayInformation.GetForCurrentView`, `DisplayInformation.GetAdvancedColorInfo`, and look at `AdvancedColorInfo.CurrentAdvancedColorKind`?

walbourn commented:
The "official" DX12 HDR sample for Desktop doesn't use ``MonitorFromWindow`` but checks for the 'best' window intersection with desktop coords directly to match an adapter.

https://github.com/microsoft/DirectX-Graphics-Samples/blob/master/Samples/Desktop/D3D12HDR/src/D3D12HDR.cpp

For UWP it does the same thing:

https://github.com/microsoft/DirectX-Graphics-Samples/blob/master/Samples/UWP/D3D12HDR/src/D3D12HDR.cpp

I should also be handling ``WM_DISPLAYCHANGE`` for Win32 and ``OnDisplayContentsInvalidated`` for UWP.

jpark37 commented:
I moved the color space update to happen in response to WM_DISPLAYCHANGE, and copied the intersection test out of D3D12HDR.

Also added a null check for m_swapChain because it doesn't make sense to test for null early for the function, and assume non-null later in the function.

walbourn commented:
Thanks. I'm checking into if the Win32 official sample should be using MonitorFromWindow instead, but this works too :)

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/DeviceResources.cpp:
```diff
@@ -398,6 +398,12 @@ bool DeviceResources::WindowSizeChanged(int width, int height)
 }
 
+// This method is called when the Win32 display changes
+void DeviceResources::DisplayChanged()
```
If you look at the linked PRs, I decided to make ``UpdateColorSpace`` a public method. That method first checks to see if the DXGI factory ``IsCurrent`` and if it is not, it recreates it.

This simplified Present() where it just calls ``UpdateColorSpace`` if ``IsCurrent`` return false. It also made it easier to handle ``WM_DISPLAYCHANGE`` without adding yet another 'pass-through' public method.

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/DeviceResources.cpp:
```diff
@@ -540,6 +546,17 @@ void DeviceResources::GetHardwareAdapter(IDXGIAdapter1** ppAdapter)
+{
+    return std::max(0, std::min(ax2, bx2) - std::max(ax1, bx1)) * std::max(0, std::min(ay2, by2) - std::max(ay1, by1));
+}
+
```
Move this up to the anonymous namespace at the top of the file.

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/DeviceResources.cpp:
```diff
@@ -550,11 +567,52 @@ void DeviceResources::UpdateColorSpace()
     {
-        ComPtr<IDXGIOutput> output;
-        if (SUCCEEDED(m_swapChain->GetContainingOutput(output.GetAddressOf())))
+        if (!m_dxgiFactory->IsCurrent())
```
I did this at the very top of the UpdateColorSpace function:

```
if (!m_dxgiFactory)
    return;

if (!m_dxgiFactory->IsCurrent())
{
    // Output information is cached on the DXGI Factory. If it is stale we need to create a new factory.
   CreateFactory();
}
```

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/DeviceResources.h:
```diff
@@ -34,6 +34,7 @@ namespace DX
         void CreateWindowSizeDependentResources();
         void SetWindow(HWND window, int width, int height);
         bool WindowSizeChanged(int width, int height);
+        void DisplayChanged();
```
Moved ``UpdateColorSpace`` to public instead of adding a new public method.

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/SimpleHDR_PC.cpp:
```diff
@@ -789,6 +789,11 @@ void Sample::OnWindowSizeChanged(int width, int height)
 
+void Sample::OnDisplayChanged()
+{
+	m_deviceResources->DisplayChanged();
```
This becomes ``m_deviceResources->UpdateColorSpace();``

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/DeviceResources.cpp:
```diff
@@ -550,11 +567,52 @@ void DeviceResources::UpdateColorSpace()
+                const int intersectArea = ComputeIntersectionArea(ax1, ay1, ax2, ay2, bx1, by1, bx2, by2);
+                if (intersectArea > bestIntersectArea)
+                {
+                    bestOutput = std::move(output);
```
Since this is a ``ComPtr`` a better choice is ``bestOutput.Swap(output);``.

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/Main.cpp:
```diff
@@ -182,6 +182,13 @@ LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
         }
         break;
 
+    case WM_DISPLAYCHANGE:
```
FWIW, I positioned this just before ``WM_MOVE``

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/DeviceResources.cpp:
```diff
@@ -550,11 +567,52 @@ void DeviceResources::UpdateColorSpace()
+
+        // Get the retangle bounds of the app window
+        RECT windowBounds;
+        GetWindowRect(m_window, &windowBounds);
```
Technically ``GetWindowRect`` returns a BOOL for success or failure. 

How about:

```
if (!GetWindowRect(m_window, &windowBounds))
    throw DX::com_exception(HRESULT_FROM_WIN32(GetLastError()));
````

> Note in the modern template, I'd actually use this mouthful:  ``throw .std::system_error(std::error_code(static_cast<int>(GetLastError()), std::system_category()), "GetWindowRect");``. These older templates didn't use that pattern or include ``<system_error>``.

> For true cases of "I don't ever care about the return" you can use ``(void)GetWindowRect(...);`` but that upsets some compiler tidy warnings. In the modern template, I'd have used ``std::ignore = GetWindowRect(...``, but this older template doesn't include ``<tuple>``.

walbourn commented on PCSamples/Graphics/SimpleHDR_PC/DeviceResources.cpp:
```diff
@@ -550,11 +567,52 @@ void DeviceResources::UpdateColorSpace()
+                // Get the rectangle bounds of current output
+                DXGI_OUTPUT_DESC desc;
+                ThrowIfFailed(output->GetDesc(&desc));
+                RECT r = desc.DesktopCoordinates;
```
For clang tidiness sake, the parameters to ``ComputeIntersectionArea``, it's return value, and all these locals should be ``long`` instead of ``int`` since the underlying type of ``RECT`` members is actually ``long``. MSVC considers them the same, but they are not strictly speaking the same type.

walbourn reviewed (changes requested):
Thanks again for your contribution!

If you don't have the time or interest in all the cleanup feedback, I'm happy to deal with it later. I'm going to finish up the linked PRs first and then circle back to this one in any case. Plus I need to work out how to test this myself since my HDR monitor is at the office...

jpark37 commented:
Copied in changes as they are for d3d11game_win32_dr, which I think covers all the requests. Also had to add `<system_error>` to `pch.h` to fix compiler errors.

walbourn commented:
Thanks. I'll also need to update SimpleHDR_PC12, SimpleHDR_UWP, and SimpleHDR_UWP12 for the appropriate changes, but I can do that.

jpark37 commented:
Sounds good; I'll let you take care of the rest.

I'd be surprised if MS wouldn't let you borrow an HDR monitor considering you have a need for one, like testing these changes. I dunno how close you guys are to being able to go back in office.

walbourn commented:
Due to our release mechanism, I ended up doing this via a different path but here is the commit with all your changes, plus also the other sample updates.

If you can try out SimpleHDR_PC, SimpleHDR_PC12, SimpleHDR_UWP, SimpleHDR_UWP12, and SimplePBR12 for UWP that would be helpful. Otherwise I can take it on faith :)

https://github.com/microsoft/Xbox-ATG-Samples/commit/187562b015d83679caf1e1ff93bca9bfc4fe7942

jpark37 commented:
Verified all five seemed to respond correctly to toggling HDR in settings. Thanks!

walbourn commented:
Much appreciated! Submitted all my PRs so my test suites and templates should be in good shape too.
